### ISSUE TITLE
Measure-generic structs wrongly can't be treated as unmanaged

### ISSUE THREAD

**Issue by kbattocchi** [Microsoft/visualfsharp#655]
This issue is described [here](http://fslang.uservoice.com/forums/245727-f-language/suggestions/5674080-allow-units-of-measure-on-structs-without-making-t).  Given this type definition:

```fsharp
[<Struct>] 
type vec3<[<Measure>] 'u> = 
   val mutable X: float<'u> 
   val mutable Y: float<'u> 
   val mutable Z: float<'u>
```

the F# compiler won't allow `vec3<_>` to satisfy the unmanaged type constraint because it’s “generic”, even though it isn’t really generic in the CLR sense (since measure type arguments are erased at runtime):

```fsharp
 // error FS0001: A generic construct requires that the type 'vec3<1>' is an unmanaged type
 let ptr = NativeInterop.NativePtr.ofNativeInt<vec3<1>> 0n
```

**Comment by zpodlovics** [comment]
Similar but non measure generic case for struct with measurement fields: #276 which was fixed in #289

```fsharp
open System.Runtime.InteropServices

[<Measure>]
type Version

[<Struct>]
type S =
  val X: int

[<Struct>]
type SM =
  val X: int<Version> 

[<Struct>]
type SMG<[<Measure>] 'u> =
  val X: int<'u> 

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
type SS =
  val X: int

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
type SMS =
  val X: int<Version> 

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
type SMGS<[<Measure>] 'u> =
  val X: int<'u> 

let f (x : 'T when 'T: unmanaged and 'T: struct) = printfn "%A" x

f (S())
f (SM())
//f (SMG()) // FS0001: A generic construct requires that the type 'SMG<'u>' is an unmanaged type
//f (SMG<Version>()) // FS0001: A generic construct requires that the type 'SMG<Version>' is an unmanaged type
f (SS())
f (SMS())
//f (SMGS()) // FS0001: A generic construct requires that the type 'SMGS<'u>' is an unmanaged type
//f (SMGS<Version>()) // FS0001: A generic construct requires that the type 'SMGS<Version>' is an unmanaged type

printf "is SMG<Version> a generic type: %A\n" (typeof<SMG<Version>>.IsGenericType)
printf "is SMG<Version> a value type: %A\n" (typeof<SMG<Version>>.IsValueType)
```

Note if you use it for native interop:
Blittable types are a subset of unmanaged types where the type have same managed and unmanaged representation.

Ecma 355 II.10.1.2 Type layout attributes

"[Rationale: The default auto layout should provide the best layout for the platform on which the code
is executing. sequential layout is intended to instruct the CLI to match layout rules commonly
followed by languages like C and C++ on an individual platform, where this is possible while still
guaranteeing verifiable layout. explicit layout allows the CIL generator to specify the precise layout
semantics. end rationale]"

**Comment by dsyme** [comment]
I think this counts as a feature request, please add it to http://fslang.uservoice.com/ and link back here. Thanks!

**Comment by kbattocchi** [comment]
@dsyme Actually it was already reported on user voice, but closed - see [here](http://fslang.uservoice.com/forums/245727-f-language/suggestions/5674080-allow-units-of-measure-on-structs-without-making-t).

**Comment by zpodlovics** [comment]
@dsyme  was right about Allow units of measure on structs without making them generic, the measure generic structure is still a non generic value type, you can check it yourself. My example code will show the following message:

```
is SMG<Version> a generic type: false
is SMG<Version> a value type: true
```

A note about your user voice comment: 
"This has the unfortunate side-effect of making the struct generic (and thus not blittable, and increasing the size by 8 bytes)". 
Please see my earlier comment about unmanaged types vs blittable types, because not every unmanaged type are blittable.

Is there any defined semantics available for measure generic structs? Something weird going on, because the bytecode representation shows no difference (except the names), but the struct size increased.

type S
```
  .class nested public sequential ansi sealed serializable S
  	extends [mscorlib]System.ValueType
  {
    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.StructAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoEqualityAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoComparisonAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) =  (01 00 03 00 00 00 00 00 ) // ........

    .field  assembly  int32 X@

    // method line 4
    .method public hidebysig specialname 
           instance default int32 get_X ()  cil managed 
    {
        // Method begins at RVA 0x208c
	// Code size 7 (0x7)
	.maxstack 8
	IL_0000:  ldarg.0 
	IL_0001:  ldfld int32 Structgen/S::X@
	IL_0006:  ret 
    } // end of method S::get_X

	.property instance int32 X ()
	{
		.custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags, int32) =  (01 00 04 00 00 00 00 00 00 00 00 00 ) // ............

		.get instance default int32 Structgen/S::get_X () 
	}
  } // end of class S
```

type SM
```
  .class nested public sequential ansi sealed serializable SM
  	extends [mscorlib]System.ValueType
  {
    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.StructAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoEqualityAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoComparisonAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) =  (01 00 03 00 00 00 00 00 ) // ........

    .field  assembly  int32 X@

    // method line 5
    .method public hidebysig specialname 
           instance default int32 get_X ()  cil managed 
    {
        // Method begins at RVA 0x2094
	// Code size 7 (0x7)
	.maxstack 8
	IL_0000:  ldarg.0 
	IL_0001:  ldfld int32 Structgen/SM::X@
	IL_0006:  ret 
    } // end of method SM::get_X

	.property instance int32 X ()
	{
		.custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags, int32) =  (01 00 04 00 00 00 00 00 00 00 00 00 ) // ............

		.get instance default int32 Structgen/SM::get_X () 
	}
  } // end of class SM
```

type SMG
```
  .class nested public sequential ansi sealed serializable SMG
  	extends [mscorlib]System.ValueType
  {
    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.StructAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoEqualityAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.NoComparisonAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) =  (01 00 03 00 00 00 00 00 ) // ........

    .field  assembly  int32 X@

    // method line 6
    .method public hidebysig specialname 
           instance default int32 get_X ()  cil managed 
    {
        // Method begins at RVA 0x209c
	// Code size 7 (0x7)
	.maxstack 8
	IL_0000:  ldarg.0 
	IL_0001:  ldfld int32 Structgen/SMG::X@
	IL_0006:  ret 
    } // end of method SMG::get_X

	.property instance int32 X ()
	{
		.custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags, int32) =  (01 00 04 00 00 00 00 00 00 00 00 00 ) // ............

		.get instance default int32 Structgen/SMG::get_X () 
	}
  } // end of class SMG
```

unified diff between S and SMG
```diff
--- S.il	2015-10-05 17:45:21.402557379 +0200
+++ SMG.il	2015-10-05 17:45:40.866749017 +0200
@@ -1,4 +1,4 @@
-  .class nested public sequential ansi sealed serializable S
+  .class nested public sequential ansi sealed serializable SMG
   	extends [mscorlib]System.ValueType
   {
     .custom instance void class [FSharp.Core]Microsoft.FSharp.Core.StructAttribute::'.ctor'() =  (01 00 00 00 ) // ....
@@ -11,22 +11,22 @@
 
     .field  assembly  int32 X@
 
-    // method line 4
+    // method line 6
     .method public hidebysig specialname 
            instance default int32 get_X ()  cil managed 
     {
-        // Method begins at RVA 0x208c
+        // Method begins at RVA 0x209c
 	// Code size 7 (0x7)
 	.maxstack 8
 	IL_0000:  ldarg.0 
-	IL_0001:  ldfld int32 Structgen/S::X@
+	IL_0001:  ldfld int32 Structgen/SMG::X@
 	IL_0006:  ret 
-    } // end of method S::get_X
+    } // end of method SMG::get_X
 
 	.property instance int32 X ()
 	{
 		.custom instance void class [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::'.ctor'(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags, int32) =  (01 00 04 00 00 00 00 00 00 00 00 00 ) // ............
 
-		.get instance default int32 Structgen/S::get_X () 
+		.get instance default int32 Structgen/SMG::get_X () 
 	}
-  } // end of class S
+  } // end of class SMG

```

```fsharp
open System.Runtime.InteropServices

#nowarn "9"

[<Measure>]
type Version

[<Struct>]
[<NoEquality;NoComparison>]
type S =
  val X: int

[<Struct>]
[<NoEquality;NoComparison>]
type SM =
  val X: int<Version> 

[<Struct>]
[<NoEquality;NoComparison>]
type SMG<[<Measure>] 'u> =
  val X: int<'u> 

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
[<NoEquality;NoComparison>]
type SS =
  val X: int

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
[<NoEquality;NoComparison>]
type SMS =
  val X: int<Version> 

[<Struct; StructLayout(LayoutKind.Sequential, Pack=1)>]
[<NoEquality;NoComparison>]
type SMGS<[<Measure>] 'u> =
  val X: int<'u> 

let f (x : 'T when 'T: unmanaged and 'T: struct) = printfn "%A" x

f (S())
f (SM())
//f (SMG()) // FS0001: A generic construct requires that the type 'SMG<'u>' is an unmanaged type
//f (SMG<Version>()) // FS0001: A generic construct requires that the type 'SMG<Version>' is an unmanaged type
f (SS())
f (SMS())
//f (SMGS()) // FS0001: A generic construct requires that the type 'SMGS<'u>' is an unmanaged type
//f (SMGS<Version>()) // FS0001: A generic construct requires that the type 'SMGS<Version>' is an unmanaged type

printf "is 64 bit process: %A\n" (System.Environment.Is64BitProcess)
printf "pointer size in bits: %A\n" (System.IntPtr.Size)

printf "is SMG<Version> a generic type: %A\n" (typeof<SMG<Version>>.IsGenericType)
printf "is SMG<Version> a value type: %A\n" (typeof<SMG<Version>>.IsValueType)

printf "sizeof<S>: %A\n" (sizeof<S>)
printf "sizeof<SM>: %A\n" (sizeof<SM>)
printf "sizeof<SMG<Version>>: %A\n" (sizeof<SMG<Version>>)

printf "sizeof<SS>: %A\n" (sizeof<SS>)
printf "sizeof<SMS>: %A\n" (sizeof<SMS>)
printf "sizeof<SMGS<Version>: %A\n" (sizeof<SMGS<Version>>)

printf "Marshal.SizeOf(typeof<S>): %A\n" (Marshal.SizeOf(typeof<S>))
printf "Marshal.SizeOf(typeof<SM>): %A\n" (Marshal.SizeOf(typeof<SM>))
printf "Marshal.SizeOf(sizeof<SMG<Version>>): %A\n" (Marshal.SizeOf(typeof<SMG<Version>>))

printf "Marshal.SizeOf(typeof<SS>): %A\n" (Marshal.SizeOf(typeof<SS>))
printf "Marshal.SizeOf(typeof<SMS>): %A\n" (Marshal.SizeOf(typeof<SMS>))
printf "Marshal.SizeOf(typeof<SMGS<Version>): %A\n" (Marshal.SizeOf(typeof<SMGS<Version>>))

```

Something weird going on because the compiled version (exe) does not show the same result when it executed from fsi.

Compiled version:
```
Structgen+S
Structgen+SM
Structgen+SS
Structgen+SMS
is 64 bit process: true
pointer size in bits: 8
is SMG<Version> a generic type: false
is SMG<Version> a value type: true
sizeof<S>: 4
sizeof<SM>: 4
sizeof<SMG<Version>>: 4
sizeof<SS>: 4
sizeof<SMS>: 4
sizeof<SMGS<Version>: 4
Marshal.SizeOf(typeof<S>): 4
Marshal.SizeOf(typeof<SM>): 4
Marshal.SizeOf(sizeof<SMG<Version>>): 4
Marshal.SizeOf(typeof<SS>): 4
Marshal.SizeOf(typeof<SMS>): 4
Marshal.SizeOf(typeof<SMGS<Version>): 4
```

fsi session (fsharpi --load:structgen.fsx)

```
FSI_0002.Structgen+S
FSI_0002.Structgen+SM
FSI_0002.Structgen+SS
FSI_0002.Structgen+SMS
is 64 bit process: true
pointer size in bits: 8
is SMG<Version> a generic type: false
is SMG<Version> a value type: true
sizeof<S>: 4
sizeof<SM>: 4
sizeof<SMG<Version>>: 8
sizeof<SS>: 4
sizeof<SMS>: 4
sizeof<SMGS<Version>: 8
Marshal.SizeOf(typeof<S>): 4
Marshal.SizeOf(typeof<SM>): 4
Marshal.SizeOf(sizeof<SMG<Version>>): 8
Marshal.SizeOf(typeof<SS>): 4
Marshal.SizeOf(typeof<SMS>): 4
Marshal.SizeOf(typeof<SMGS<Version>): 8
```

Update: something weird going on within the fsi session compared to the compiled execution, because the struct size is different (8 byte vs 4 byte on compiled case) on the measure generic struct case. Maybe a value type compilation / representation bug?

**Comment by kbattocchi** [comment]
@zpodlovics See issue #654, which is being fixed by #656.
